text/1011-object-level-security.rst

```diff
@@ -0,0 +1,320 @@
+rules in schemas to automatically enforce data access and modification on
+per-object level.  This is analogous to row-level security (RLS) in SQL
+systems.
+
+
+Motivation
+==========
+
+Object-level security is a powerful feature which allows universal leak-proof
+enforcement of data access policies across all uses of a database.  This is
+important for security and compliance, but would also be tremendously useful
+for backend-less applications that interact with EdgeDB via HTTP (either via
+GraphQL or EdgeQL-over-HTTP).  A secondary motivation is that this will allow
+implementation of `temporal databases <temporal>`_.
+
+
+Requirements
+============
+
+Although there is no direct dependency, this RFC requires
+`RFC 1010 <1001-global-vars.rst>`_ (globals) to be implemented to be practical.
+See `Interaction with globals`_ below for details.
+
+
+Specification
+=============
+
+This RFC proposes to add a new schema item type -- *Access Policy* -- that is
+defined in the context of an object type and specifies the rules of rewriting
+queries that refer to the enclosing object type.
+
+CREATE ACCESS POLICY
+--------------------
+
+Define a new access control policy for a given object type.
+
+Required capabilities: DDL.
+
+Synopsis::
+
+    {CREATE|ALTER} OBJECT TYPE <type-name> "{"
+        CREATE ACCESS POLICY <name>
+            [ WHEN <condition> ]
+            { PERMIT | RESTRICT }
+            { ALL | READ | WRITE | DELETE } [ ... ]
+            USING (<expr>)
+    "}"
+
+An access policy grants the permission to select, insert or delete objects
+that match the relevant policy expression.  The ``READ`` and ``DELETE``
+policies specify a *visibility* check for existing objects: when ``<expr>>``
+returns a false result for an object, such object will be silently skipped
+in any ``SELECT``, ``UPDATE`` or ``DELETE`` expression that scans it.
+A ``WRITE`` policy specifies a *validity* check for new or updated objects and
+affects ``INSERT`` and ``UPDATE`` expressions: if the proposed object violates
+``<expr>``, an error is raised immediately and the query is aborted.
+
+The optional ``<condition>`` expression is evaluated for every object
+affected by the statement and the policy is applied only if the expression
+evaluates to *true*.  It is essentially equivalent to joining ``<condition>``
+with ``<expr>`` with an ``AND`` operator.  The reason for a standalone clause
+is that it makes it easier to separate *when* a policy is applied from
+*how* a policy is applied.  Also, other access policies do not apply to
+``when`` expressions, but they *do* apply to ``using`` expressions.
+
+If at least one access rule is defined for a type, then all elements become
+invisible/immutable by default and must be made visible/mutable by at least
+one ``PERMIT`` rule.
+
+A ``PERMIT`` rule *adds* all objects, for which the check expression
+evaluates to true, to the set of visible objects.  ``PERMIT`` rules are
+combined using the ``OR`` operator.
```

Comment:
What's the intended behavior for `{}`? My inclination would be to treat it like `false`. Equivalently, that the rules are essentially combined with `any`

Reply:
`{}` evalueates to `false` like it does in `filter` clauses.